fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,444 @@
+    }
+
+    fn block_hash(&self, height: BlockHeight) -> StmResult<Option<Bytes>> {
+        let cache = self.height_data.read()?;
+        Ok(cache.get_value(height).map(|i| i.0.clone()))
+    }
+
+    fn validator_set(&self, height: BlockHeight) -> StmResult<Option<ValidatorSet>> {
+        let cache = self.height_data.read()?;
+        Ok(cache.get_value(height).map(|i| i.1.clone()))
+    }
+
+    fn all_top_down_msgs(&self) -> StmResult<Vec<CrossMsg>> {
+        let cache = self.top_down_msgs.read()?;
+        Ok(cache.values().into_iter().cloned().collect())
+    }
+}
+
+#[async_trait]
+impl ParentViewProvider for DefaultFinalityProvider {
+    async fn latest_height(&self) -> Option<BlockHeight> {
+        atomically(|| self.parent_view_data.latest_height()).await
+    }
+
+    async fn latest_nonce(&self) -> Option<Nonce> {
+        atomically(|| {
+            let top_down_msgs = self.parent_view_data.top_down_msgs.read()?;
+            Ok(top_down_msgs.upper_bound())
+        })
+        .await
+    }
+
+    async fn new_parent_view(
+        &self,
+        block_info: Option<(BlockHeight, Bytes, ValidatorSet)>,
+        mut top_down_msgs: Vec<CrossMsg>,
+    ) -> Result<(), Error> {
+        top_down_msgs.sort_unstable_by(|a, b| {
+            a.msg.nonce.cmp(&b.msg.nonce)
+        });
+
+        atomically(|| {
+            if let Some((height, hash, validator_set)) = &block_info {
+                let insert_res = self.parent_view_data.height_data.modify(|mut cache| {
+                    let r = cache.insert(*height, (hash.clone(), validator_set.clone()));
+                    (cache, r)
+                })?;
+                match insert_res {
+                    SequentialCacheInsert::Ok => {}
+                    // now the inserted height is not the next expected block height, could be a chain
+                    // reorg if the caller is behaving correctly.
+                    _ => return Ok(Err(Error::ParentReorgDetected(*height))),
+                };
+            }
+
+            if top_down_msgs.is_empty() {
+                // not processing if there are no top down msgs
+                return Ok(Ok(()));
+            }
+
+            // get the min nonce from the list of top down msgs and purge all the msgs with nonce
+            // about the min nonce in cache, as the data should be newer and more accurate.
+            let min_nonce = top_down_msgs.first().unwrap().msg.nonce;
+            self.parent_view_data.top_down_msgs.modify(|mut cache| {
+                cache.remove_key_above(min_nonce);
+
+                for msg in top_down_msgs.clone() {
+                    cache.insert(msg.msg.nonce, msg);
+                }
+                (cache, ())
+            })?;
+
+            Ok(Ok(()))
+        })
+        .await
+    }
+}
+
+#[async_trait]
+impl ParentFinalityProvider for DefaultFinalityProvider {
+    async fn last_committed_finality(&self) -> IPCParentFinality {
+        atomically(|| {
```

Comment:
I see that all these methods do is run `atomically` and nothing else. In this case it might be better to actually return `StmResult` and let the caller decide on what boundaries they want to combine these methods.

It would make it easier to see that this "provider" is an in-memory component with no immediate side effects such as no going to the Agent. If we need that, we can make it obvious in the code when it happens by adding another dependency.

Reply:
updated